node/network/statement-distribution/src/lib.rs

```diff
@@ -911,38 +1098,249 @@ async fn report_peer(
 	)).await
 }
 
-// Handle an incoming wire message. Returns a reference to a newly-stored statement
-// if we were not already aware of it, along with the corresponding relay-parent.
-//
-// This function checks the signature and ensures the statement is compatible with our
-// view. It also notifies candidate backing if the statement was previously unknown.
-#[tracing::instrument(level = "trace", skip(peer_data, ctx, active_heads, metrics), fields(subsystem = LOG_TARGET))]
-async fn handle_incoming_message<'a>(
+async fn handle_incoming_message_and_circulate<'a>(
 	peer: PeerId,
-	peer_data: &mut PeerData,
-	active_heads: &'a mut HashMap<Hash, ActiveHeadData>,
-	ctx: &mut impl SubsystemContext<Message = StatementDistributionMessage>,
+	peers: &mut HashMap<PeerId, PeerData>,
 	message: protocol_v1::StatementDistributionMessage,
+	active_heads: &'a mut HashMap<Hash, ActiveHeadData>,
+	ctx: &mut impl SubsystemContext,
+	req_sender: &mpsc::Sender<RequesterMessage>,
 	metrics: &Metrics,
-) -> Option<(Hash, &'a StoredStatement)> {
-	let (relay_parent, statement) = match message {
-		protocol_v1::StatementDistributionMessage::Statement(r, s) => (r, s),
-	};
+) {
+	let relay_parent = message.get_metadata().relay_parent;
+	let result = retrieve_statement_from_message(
+		peer,
+		message,
+		active_heads,
+		ctx,
+		req_sender,
+	).await;
+
+	if let Some((active_head, statement)) = result {
+		handle_incoming_statement_and_circulate(
+			peer,
+			peers,
+			active_head,
+			ctx,
+			relay_parent,
+			statement,
+			metrics,
+		).await;
+	}
+}
 
-	let active_head = match active_heads.get_mut(&relay_parent) {
+/// If message contains a statement, then retrieve it, otherwise fork task to fetch it.
+///
+/// This function will also return `None` if the message did not pass some basic checks, in that
+/// case no statement will be requested, on the flipside you get `ActiveHeadData` in addition to
+/// your statement.
+///
+/// If the message was large, but the result has been fetched already that one is returned.
+async fn retrieve_statement_from_message<'a>(
+	peer: PeerId,
+	message: protocol_v1::StatementDistributionMessage,
+	active_heads: &'a mut HashMap<Hash, ActiveHeadData>,
+	ctx: &mut impl SubsystemContext,
+	req_sender: &mpsc::Sender<RequesterMessage>,
+) -> Option<(&'a mut ActiveHeadData, SignedFullStatement)> {
+
+	let metadata = message.get_metadata();
+
+	let active_head = match active_heads.get_mut(&metadata.relay_parent) {
 		Some(h) => h,
 		None => {
 			tracing::debug!(
 				target: LOG_TARGET,
-				?peer,
-				?relay_parent,
-				"Unknown (or outdated) relay parent"
+				requested_relay_parent = %metadata.relay_parent,
+				"our view out-of-sync with active heads; head not found",
 			);
 			report_peer(ctx, peer, COST_UNEXPECTED_STATEMENT).await;
-			return None;
+			return None
+		}
+	};
+
+	match message {
+		protocol_v1::StatementDistributionMessage::Statement(_, statement) =>
+			return Some((active_head, statement)),
```

Comment:
If a peer sends us a statement here which is bigger than expected, i.e. what would normally be a large statement, should we report the peer with some mild cost?

Reply:
Not sure this is necessary. After this PR we can set notification sizes back to sane values, so a validator will not be able to send large notifications anyway.